lightning/src/blinded_path/payment.rs

```diff
@@ -276,9 +293,17 @@ pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
 ) -> Result<Vec<BlindedHop>, secp256k1::Error> {
 	let pks = intermediate_nodes.iter().map(|node| &node.node_id)
 		.chain(core::iter::once(&payee_node_id));
-	let tlvs = intermediate_nodes.iter().map(|node| BlindedPaymentTlvsRef::Forward(&node.tlvs))
-		.chain(core::iter::once(BlindedPaymentTlvsRef::Receive(&payee_tlvs)));
-	utils::construct_blinded_hops(secp_ctx, pks, tlvs, session_priv)
+	let tlvs = intermediate_nodes.iter().map(|node| BlindedPaymentTlvs::Forward(node.tlvs.clone()))
+		.chain(core::iter::once(BlindedPaymentTlvs::Receive(payee_tlvs)));
+
+	let max_length = tlvs.clone()
+		.map(|tlv| tlv.serialized_length())
+		.max()
+		.unwrap_or(0);
+
+	let length_tlvs = tlvs.map(|tlv| tlv.pad_tlvs(max_length));
+
+	utils::construct_blinded_hops(secp_ctx, pks, length_tlvs, session_priv)
```

Comment:
Something worth exploring would be to keep this using `BlindedPaymentTlvsRef` and add `BlindedPaymentTlvsMut` to hold mutable references. Then use the first to compute `max_length` and then the the second for `length_tlvs`. I _think_ that would avoid cloning, but I haven't attempted it.

Reply:
Yeah, you're right! I forgot about `pks`. 🙁 

However, I think we can get around this, but it requires a bit more refactoring.  We can have `construct_blinded_hops` take a single iterator (instead of two) yielding `(PublicKey, Option<W>)` where `W` implements `Writeable`. Using a copy of `PublicKey` instead of a reference is fine.

Here's a [branch](https://github.com/jkczyz/rust-lightning/commits/2024-08-padding-using-one-iterator/) where I was able to get it to work. It's a bit ugly in that `construct_keys_callback` is also used in `messenger.rs` for forming the onion, and there we don't pass any TLVs in. It would be better if we could refactor the code such that we don't need to pair those pubkeys with `None`. Also, for the `BlindedPath` construction, passing the TLVs into the macro and then having it pass them to the callback is a bit odd. Maybe you can find a cleaner way to achieve this?

Reply:
Ok, I ended updating the branch to clean up the approach. I also made a clean [branch](https://github.com/jkczyz/rust-lightning/commits/2024-08-blinded-path-utils-refactor/), which is slightly different and breaks the change up into smaller commits to show what is going on. I'll probably open up for review, but feel free to rebase on it.

Reply:
Thinking on this a bit more, the refactor is probably not entirely necessary. We could have just used a `for` loop to pad the TLVs before constructing the iterators. But it is a nice clean-up and would make the code neater. It also would be necessary when adding dummy hops to avoid allocations assuming they are added in the creation functions rather than the `Router` implementation.